File: solutions/haskell/collatz-conjecture/1/src/CollatzConjecture.hs
module CollatzConjecture (collatz) where

collatz :: Integer -> Maybe Integer
collatz n
 | n <= 0 = Nothing
 | otherwise = Just (countSteps n 0)

countSteps :: Integer -> Integer -> Integer
countSteps n steps 
  | n == 1 = steps
  | even n = countSteps (n `div` 2) next
  | otherwise = countSteps (3 * n + 1) next
  where next = steps + 1

File: solutions/cpp/collatz-conjecture/1/collatz_conjecture.cpp
#include "collatz_conjecture.h"
#include <stdexcept>

namespace collatz_conjecture {
    int steps(int num) {
        if(num <= 0) throw std::domain_error("Uh-oh");
        return countSteps(num, 0);
    }

    int countSteps(int num, int steps) {
        if(num == 1) return steps;
        if(num % 2 == 0)
            return countSteps(num / 2, steps + 1);
        else
            return countSteps(3 * num + 1, steps + 1);
    }
}  // namespace collatz_conjecture
